### ISSUE TITLE
A way to make eglot more manageable

### ISSUE THREAD

**Issue by guskovd** [joaotavora/eglot#464]
First of all, thanks for eglot-ignored-server-capabilites

In lsp.el did not even guess to do this.

But by and large, using eglot.el and lsp.el is equally difficult.

For example, I do not want to use the built-in flymake-checker. To do this, I set the hook:

```lisp
(progn
  (eglot-ensure)
  (flymake-mode -1)
)
```

But after several operations with the text, flymake-mode turns on again!

I don’t understand what kind of hook starts it, and why this happens.

--

Why not leave only the lsp-server by default and not allow users to set hooks themselves on those events that they consider necessary? This is emacs-way

**Comment by joaotavora** [comment]
By default eglot turns on all bells and whistles, but there `eglot-stay-out-of`.  I'll check if it can be used to completely turn off flymake completely, I'm not sure.

**Comment by guskovd** [comment]
> By default eglot turns on all bells and whistles, but there `eglot-stay-out-of`. I'll check if it can be used to completely turn off flymake completely, I'm not sure.

Thank. This is a useful option.
But flymake all the same starts automatically when you call eglot-ensure.
I think this is not very correct

**Comment by joaotavora** [comment]
I thinkt he way is to add the `flymake-mode -1` thing to `eglot-managed-mode-hook`.  Can you try that?

**Comment by guskovd** [comment]
Thanks! This is my config for rust-mode:

```
(defun my-eglot-managed-mode-hook ()
  "My eglot hook."
  (flymake-mode -1)
  (eldoc-mode -1)
  ;; (remove-hook 'after-change-functions 'eglot--after-change t)
  (remove-hook 'before-change-functions 'eglot--before-change t)
  (remove-hook 'kill-buffer-hook #'eglot--managed-mode-off t)
  (remove-hook 'kill-buffer-hook 'eglot--signal-textDocument/didClose t)
  (remove-hook 'before-revert-hook 'eglot--signal-textDocument/didClose t)
  (remove-hook 'after-revert-hook 'eglot--after-revert-hook t)
  (remove-hook 'before-save-hook 'eglot--signal-textDocument/willSave t)
  (remove-hook 'after-save-hook 'eglot--signal-textDocument/didSave t)
  ;; (remove-hook 'xref-backend-functions 'eglot-xref-backend t)
  ;; (remove-hook 'completion-at-point-functions #'eglot-completion-at-point t)
  (remove-hook 'change-major-mode-hook #'eglot--managed-mode-off t)
  (remove-hook 'post-self-insert-hook 'eglot--post-self-insert-hook t)
  (remove-hook 'pre-command-hook 'eglot--pre-command-hook t)

  )

(add-hook 'eglot-managed-mode-hook 'my-eglot-managed-mode-hook)
```

**Comment by guskovd** [comment]
But I'm still not sure if this is the right approach.
In my opinion, many people use eglot rather than lsp.el because it is more subtle in its configuration.

I am convinced that if you make the eglot - managed-mode function more customizable, then it will attract even more users.

For example, you can make a function
```lisp
(defun eglot-auto-configure() 
 ...
)
```
and option

```lisp
(defvar eglot-auto-configure-enabled t)
```

**Comment by joaotavora** [comment]
> Thanks! This is my config for rust-mode:

Whooo, if seems you basically disable all of Eglot there!  No, you shouldn't do that.   You shouldn't every type an `eglot--` symbol (with two dashes).  I don't want you to do that.  Or rather, you _can_ do it, of course.  I won't stop you.  But I won't be able to provide suppport if you do.

Anyway, a request to not have Flymake.  I think that is resolved.

> In my opinion, many people use eglot rather than lsp.el because it is more subtle in its configuration.

In my opinion, it is rather because it works out of the box with just Emacs and nothing more.

Why don't you start by explaining what you use Eglot for?

**Comment by guskovd** [comment]
Thanks for the quick response!

I think my problem is resolved.

Answering your question, I only need the auto-complete && find-definition functions. Until recently from eglot (or lsp.el), I used racer.el to autocomplete, flycheck - to check synatxis, yasnippet to insert snippets - everything is as usual in the old school emacs ...

It seems that the rust community has opted for rust-analyzer (working through lsp) and does not support racer well.

When I had to rewrite the configs, I was surprised that the libraries implementing lsp are so slow. Essentially, they turn emacs into a slow-slow IDE. I don't need it at all
